.github/workflows/integration-cloud.yml

```
@@ -13,20 +13,58 @@ concurrency:
   cancel-in-progress: false
 
 jobs:
+  setup-environment:
+    runs-on: ubuntu-latest
+    outputs:
+      python-cache-key: ${{ steps.cache-python.outputs.cache-hit }}
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Set up Python
+        id: setup-python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+
+      - name: Cache dependencies
+        id: cache-python
+        uses: actions/cache@v3
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+
+      - name: Install dependencies
+        if: steps.cache-python.outputs.cache-hit != 'true'
+        run: |
+          pip install -e .[dev]
+
+      - name: Upload Python environment
+        uses: actions/upload-artifact@v3
+        with:
+          name: python-environment
+          path: ~/.cache/pip
+
   integration-cloud:
     runs-on: ubuntu-latest
+    needs: setup-environment
     steps:
       - uses: actions/checkout@v4
+
       - name: Set up Python
         uses: actions/setup-python@v5
         with:
-          python-version: "3.9"
-          cache: "pip"
-          cache-dependency-path: |
-            pyproject.toml
+          python-version: "3.10"
+
+      - name: Download Python environment
+        uses: actions/download-artifact@v3
+        with:
+          name: python-environment
+          path: ~/.cache/pip
```

Use actions/cache + `restore-key` instead?

---

Also, outdated version:
```suggestion
      - name: Download Python environment
        uses: actions/download-artifact@v4
        with:
          name: python-environment
          path: ~/.cache/pip
```